nucleus/deploy/nucleus_integration.py

```diff
@@ -0,0 +1,156 @@
+        )
+        prediction_items = []
+        for s3url, dataset_item in self.s3url_to_dataset_map.items():
+            item_link = self.hmi_async_job.responses[s3url]
+            print(f"item_link={item_link}")
+            # e.g. s3://scale-ml/tmp/hosted-model-inference-outputs/a224499e-50ac-4b08-ad0c-c18e74c14184.pkl
+            kwargs = {
+                "transport_params": {
+                    "session": Session(profile_name="ml-worker")
+                }
+            }
+
+            with smart_open.open(item_link, "rb", **kwargs) as bundle_pkl:
+                inference_result = cloudpickle.load(bundle_pkl)
+                ref_id = dataset_item.reference_id
+                for box in inference_result:
+                    # TODO assuming box is a list of (x, y, w, h, label). This is almost certainly not the case.
+                    # We will have to use a ModelEndpoint/ModelBundle that returns boxes in this format.
+                    # Also, label is probably returned as an integer instead of a label that makes semantic sense
+                    pred_item = nucleus.BoxPrediction(
+                        label=box["label"],
+                        x=box["left"],
+                        y=box["top"],
+                        width=box["width"],
+                        height=box["height"],
+                        reference_id=ref_id,
+                    )
+                    prediction_items.append(pred_item)
+
+        job = model_run.predict(prediction_items, asynchronous=True)
+        job.sleep_until_complete()
+        job.errors()
+
+
+def create_nucleus_dataset_inference_run(
+    hmi_endpoint: ModelEndpoint, nucleus_client, dataset: Dataset
+):
+    """
+    Returns a NucleusDatasetInferenceRun, client will need to periodically call poll on this in order to upload
+    """
+    s3urls, s3url_to_dataset_map = _nucleus_ds_to_s3url_list(dataset)
+    async_job = hmi_endpoint.predict(s3urls)
+    return NucleusDatasetInferenceRun(
+        async_job, nucleus_client, s3url_to_dataset_map, dataset
+    )
+
+
+def _nucleus_ds_to_s3url_list(
+    dataset: Dataset,
+) -> Tuple[List[str], Dict[str, DatasetItem]]:
+    # TODO I'm not sure if dataset items are necessarily s3URLs. Does this matter?
+    # TODO support lidar point clouds
+    if len(dataset.items) == 0:
+        logger.warning("Passed a dataset of length 0")
+        return [], {}  # TODO return type?
+    dataset_item_type = dataset.items[0].type
+    if not all(data.type == dataset_item_type for data in dataset.items):
+        logger.warning("Dataset has multiple item types")
+        raise Exception  # TODO (code style) too broad exception
+
+    s3url_to_dataset_map = {}
+    # Do we need to keep track of nucleus ids?
+    if dataset_item_type == DatasetItemType.IMAGE:
+        s3Urls = [
+            data.image_location
+            for data in dataset.items
+            if data.image_location is not None
```

Comment:
This is going to be the customer location of the URL. If they are doing hosted model inference, I believe we want to run on the processed_url, which isn't currently exported as part of the dataset item.

Reply:
Yeah this is true. I think the longer term plan is to have Nucleus itself call HMI, which makes this file unnecessary once that's built out

Reply:
Ok so I guess we need to do some work on our (Scale's) side to expose this field? If so, I can create a ticket for the ML infra team to tackle this (unless Nucleus was planning on doing this anyway).